### PR TITLE
Syntax folding for diff/patch files

### PR DESCRIPTION
**Pull request by ghost**
The syntax file for the filetype "diff" (not the mode) does currently not support folding of hunks/files. Below a draft for adding this functionality.

The folding is not 100% accurate as the "diff" filetype does cover lots of different flavours of diff or patch syntax.
The implementation below works for unified format (and thus for derivative of it, e.g. gitdiff, hgdiff, ...)

The implementation adds two syntax-item (`diffFoldFile`, `diffFoldHunk`) whose sole purpose is to provide folding. I would have liked to reuse the `diffFile` syntax item as start-pattern for the region, however, Vim does not seem to support this. Therefore, I combined the most often used one into a new regex.

This regex is not perfect, e.g. it does not contain the "^Only..." parts which are therefore folded as part of a previous hunk (if any) and not as an individual file. Adding this seems like a mess, especially because there are also a tone of translations to consider. Thus, so far only unified syntax (and a little more) is supported.

During testing I experienced the problem that in a file-fold only included the first hunk and that further hunks were not folded at all. I solved this problem with trail-and-error and derived at syncing. There are two sync way:

1. `syntax sync minlines=2 maxlines=2` (and any derivatives of it with minlines>=2 works and maxlines can be ommitted)
2. `syntax sync match diffFoldFileSync grouphere diffFoldFile '^\(...\)` (same regex as for diffFoldFile)

I don't understand any of it, except that I can see that they work. I believe "sync match" is saner as it will find the corresponding file-fold. If someone wants to shed some light on it, I'll be happy to learn something new.

It would be nice to bring this into a state such that this functionality (although not perfect for all diff flavours) can make it into the runtime files. Thanks.

P.S.: For anybody stumbling upon this in the future. Instead of using syntax folding, 'foldexpr' can be used (which has similar flaws when it comes to non unified formats):

```
" Extracted form https://github.com/sgeb/vim-diff-fold
function! DiffFoldLevel()
    let l:line=getline(v:lnum)

    if l:line =~# '^\(diff\|Index\)'     " file
        return '>1'
    elseif l:line =~# '^\(@@\|\d\)'  " hunk
        return '>2'
    elseif l:line =~# '^\*\*\* \d\+,\d\+ \*\*\*\*$' " context: file1
        return '>2'
    elseif l:line =~# '^--- \d\+,\d\+ ----$'     " context: file2
        return '>2'
    else
        return '='
    endif
endfunction

augroup ft_diff
    au!
    autocmd FileType diff setlocal foldmethod=expr
    autocmd FileType diff setlocal foldexpr=DiffFoldLevel()
augroup END
```

### CODE REVIEW

**Comment by brammool** [comment]
I'll await comments.  Perhaps some improvements can be made?

File: runtime/syntax/diff.vim [hunk 1]
```diff
@@ -371,7 +379,7 @@ hi def link diffIsA		Constant
 hi def link diffNoEOL		Constant
 hi def link diffCommon		Constant
 hi def link diffRemoved		Special
-hi def link diffChanged		PreProc
+ahi def link diffChanged		PreProc
```
Comment (jamessan) [inline runtime/syntax/diff.vim]: Typo?

**Comment by jamessan** [comment]
FWIW, this is the ftplugin I've been using for years to get diff folding:

```vim
function! DiffFoldLevel(lineno)
  let line = getline(a:lineno)
  if line =~# '^Index:'
    return '>1'
  elseif line =~# '^index '
  elseif line =~# '^===' || line =~ '^RCS file: ' || line =~ '^retrieving revision '
    let lvl = foldlevel(a:lineno - 1)
    return lvl >= 0 ? lvl : '='
  elseif line =~# '^diff'
    return getline(a:lineno - 1) =~ '^retrieving revision ' ? '=' : '>1'
  elseif line =~# '^--- ' && getline(a:lineno - 1) !~# '^diff\|^===\|^index'
    return '>1'
  elseif line =~# '^@'
    return '>2'
  elseif line =~# '^[- +\\]'
    let lvl = foldlevel(a:lineno - 1)
    return lvl >= 0 ? lvl : '='
  else
    return '0'
  endif
endf

setlocal foldmethod=expr
setlocal foldexpr=DiffFoldLevel(v:lnum)
```

**Comment by ghost** [comment]
Thanks @jamessan for sharing an extended version of the `foldexpr` function.

I'd like to point out the following difference: Adding folding support via the syntax file instead of the filetype plugin, not only provides folding support for `*.diff` and `*.patch` files but also to all filetypes which include the diff syntax, e.g. `gitcommit`.

**Pull request closed without merging**